src/invariant/BucketListIsConsistentWithDatabase.cpp

```diff
@@ -0,0 +1,104 @@
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "bucket/Bucket.h"
+#include "crypto/Hex.h"
+#include "database/Database.h"
+#include "invariant/BucketListIsConsistentWithDatabase.h"
+#include "invariant/InvariantManager.h"
+#include "ledger/AccountFrame.h"
+#include "ledger/DataFrame.h"
+#include "ledger/OfferFrame.h"
+#include "ledger/TrustFrame.h"
+#include "lib/util/format.h"
+#include "main/Application.h"
+
+namespace stellar
+{
+
+std::shared_ptr<Invariant>
+BucketListIsConsistentWithDatabase::registerInvariant(Application& app)
+{
+    return app.getInvariantManager()
+              .registerInvariant<BucketListIsConsistentWithDatabase>(
+                        app.getDatabase());
+}
+
+BucketListIsConsistentWithDatabase::BucketListIsConsistentWithDatabase(
+    Database& db)
+    : mDb{db}
+{
+}
+
+std::string
+BucketListIsConsistentWithDatabase::getName() const
+{
+    return "BucketListIsConsistentWithDatabase";
+}
+
+std::string
+BucketListIsConsistentWithDatabase::checkOnBucketApply(
+        std::shared_ptr<Bucket const> bucket,
+        uint32_t oldestLedger)
+{
+    uint64_t nAccounts = 0, nTrustLines = 0, nOffers = 0, nData = 0;
+    for (Bucket::InputIterator iter(bucket); iter; ++iter)
```

Comment:
I think we could add support for for-range for Bucket::InputIterator, should not be hard, but is it worth the time?